Set up benchmarkjunit job for structured-merge-diff

julianvmodesto opened the pull request:
It would be great to use the benchmarkjunit output to track the scalability work for structured-merge-diff.

- [pkg/benchmarkjunit](https://github.com/kubernetes/test-infra/tree/master/pkg/benchmarkjunit)
- [TestGrid demo](https://testgrid.k8s.io/sig-testing-canaries#benchmark-demo&width=5&graph-metrics=avg%20op%20duration%20(ns%2Fop)&graph-metrics=test-duration-minutes&graph-metrics=op%20count&graph-metrics=MB%2Fs&graph-metrics=alloced%20B%2Fop&graph-metrics=allocs%2Fop)

julianvmodesto commented on config/jobs/kubernetes-sigs/structured-merge-diff/structured-merge-diff-ci.yaml:
```diff
@@ -41,3 +41,31 @@ periodics:
+    path_alias: k8s.io/test-infra
+  spec:
+    containers:
+    - image: golang:latest
```
Should this be kubekins-e2e?

chases2 replied:
If this image meets your needs, it's best to use the smaller image.
Consider pinning it to a specific version of go.

julianvmodesto commented on config/jobs/kubernetes-sigs/structured-merge-diff/structured-merge-diff-ci.yaml:
```diff
@@ -41,3 +41,31 @@ periodics:
+      - ./pkg/benchmarkjunit
+      - --log-file=$(ARTIFACTS)/benchmark-log.txt
+      - --output=$(ARTIFACTS)/junit_benchmarks.xml
+      - sigs.k8s.io/structured-merge-diff/...
```
Is this module ref right?

julianvmodesto commented:
/assign @cjwagner

apelisse commented:
Thanks Julian, related to https://github.com/kubernetes/kubernetes/issues/86935

julianvmodesto commented:
Looks like @wojtek-t tried to set this up for k/k but it's blocked on https://github.com/kubernetes/kubernetes/issues/79464

- [k/k config](https://github.com/kubernetes/test-infra/blob/4225c594bbbf3082ce5c3a861d556ec3a7eda431/config/jobs/kubernetes/sig-scalability/sig-scalability-periodic-jobs.yaml#L451-L481)
- [k/k benchmark job in tesgrid](https://testgrid.k8s.io/sig-scalability-benchmarks#microbenchmarks)

Let's try this for smd.

The following options are set for the presubmit so the job is non-blocking at first and we can try it out:

```
    always_run: false
    skip_report: true
```

As described here:
https://github.com/kubernetes/test-infra/blob/76626ab4c4559f768e6519a5c5438c4b8c37b40f/prow/jobs.md#how-to-configure-new-jobs
> A useful pattern when adding new jobs is to start with always_run set to false and skip_report set to true. Test it out a few times by manually triggering, then switch always_run to true. Watch for a couple days, then switch skip_report to false.

julianvmodesto commented:
/assign @apelisse

apelisse commented:
I'm very happy that you're doing that, but I know very little about any of this to have a real "lgtm" opinion. Thanks Julian!

chases2 commented:
/lgtm
Waiting on a [local owner](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/structured-merge-diff/OWNERS) for approval

cjwagner commented on config/jobs/kubernetes-sigs/structured-merge-diff/structured-merge-diff-ci.yaml:
```diff
@@ -41,3 +41,22 @@ periodics:
+      - k8s.io/test-infra/pkg/benchmarkjunit
+      - --log-file=$(ARTIFACTS)/benchmark-log.txt
+      - --output=$(ARTIFACTS)/junit_benchmarks.xml
+      - ./...
```
This won't work, the job isn't configured to check out any code and it doesn't implicitly check out anything because it is periodically triggered (not implicitly associated with a git ref). See how `extra_refs` is used above.

julianvmodesto replied:
Oops, done: added back `extra_refs`.

cjwagner commented on config/jobs/kubernetes-sigs/structured-merge-diff/structured-merge-diff-ci.yaml:
```diff
@@ -41,3 +41,22 @@ periodics:
+      - go
+      args:
+      - run
+      - k8s.io/test-infra/pkg/benchmarkjunit
```
I don't know if this will work. I couldn't get `go run` working with go modules so we've had to check out the test-infra repo with `extra_refs` and reference by path: https://github.com/kubernetes/test-infra/blob/4225c594bbbf3082ce5c3a861d556ec3a7eda431/config/jobs/kubernetes/sig-scalability/sig-scalability-periodic-jobs.yaml#L451-L481

If you know how to make `go run` work with go modules that would be nice to get working, but I ultimately gave up.

julianvmodesto replied:
Ah right.

To address this, I've tried building a new image w/ both the benchmarkjunit binary and the go toolchain. PTAL at the change + my test.

julianvmodesto commented:
Thanks Cole!

Okay, how about this -- I've added a target to build and push a new image with the benchmarkjunit binary using golang:1.13 as the base image, so that we have the go toolchain.

I've tested this by building the image and running the image with structured-merge-diff mounted. Since go1.13 has go modules enabled by default, benchmarkjunit works successfully when it runs `go test ./...`.

```
$ bazel run pkg/benchmarkjunit:image
INFO: Analyzed target //pkg/benchmarkjunit:image (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //pkg/benchmarkjunit:image up-to-date:
  bazel-bin/pkg/benchmarkjunit/image-layer.tar
INFO: Elapsed time: 1.510s, Critical Path: 0.24s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Loaded image ID: sha256:78196c4bdf61a538e0d04dac88f5be13e417c594abcea1dfe0cfb01d60e03316
Tagging 78196c4bdf61a538e0d04dac88f5be13e417c594abcea1dfe0cfb01d60e03316 as bazel/pkg/benchmarkjunit:image

$ docker run --volume $GOPATH/src/sig.k8s.io/structured-merge-diff:/tmp/app --workdir /tmp/app bazel/pkg/benchmarkjunit:image ./...

time="2020-02-14T20:10:51Z" level=info msg="Running command [\"go\" \"test\" \"-v\" \"-run='^$'\" \"-bench=.\" \"./...\"]..."
time="2020-02-14T20:12:38Z" level=info msg="Benchmarks completed. Generating JUnit XML..."
time="2020-02-14T20:12:38Z" level=info msg="Successfully generated JUnit XML for Benchmarks."
...
```

julianvmodesto commented:
Tested with phaino:

```
$ bazel run pkg/benchmarkjunit:image
...
$ docker tag bazel/pkg/benchmarkjunit:image gcr.io/k8s-prow/benchmarkjunit:latest
...
$ bazel run //config:mkpj -- --job=ci-structured-merge-diff-benchmark > benchmark-job.yaml
...

$ bazel run //prow/cmd/phaino -- $(pwd)/benchmark-job.yaml
...
INFO[0000] Converting job into docker run command...     job=ci-structured-merge-diff-benchmark
local /path/to/sigs.k8s.io/structured-merge-diff [/home/julian/go/src/sigs.k8s.io/structured-merge-diff]:
"docker" "run" "--rm=true" \
 "--name=phaino-26122-1" \
 "--entrypoint=/app/pkg/benchmarkjunit/app.binary" \
 "-e" \
 "GOPROXY=https://proxy.golang.org" \
 "-v" \
 "/home/julian/go/src/sigs.k8s.io/structured-merge-diff:/go/src/sigs.k8s.io/structured-merge-diff" \
 "-v" \
 "/go/src/sigs.k8s.io/structured-merge-diff" \
 "-w" \
 "/go/src/sigs.k8s.io/structured-merge-diff" \
 "--label=created-by-prow=true" \
 "--label=prow.k8s.io/job=ci-structured-merge-diff-benchmark" \
 "--label=prow.k8s.io/type=periodic" \
 "--label=phaino=true" \
 "gcr.io/k8s-prow/benchmarkjunit:latest" \
 "--log-file=$(ARTIFACTS)/benchmark-log.txt" \
 "--output=$(ARTIFACTS)/junit_benchmarks.xml" \
 "./..."
INFO[0001] Starting job...                               job=ci-structured-merge-diff-benchmark
INFO[0001] Waiting for job to finish...                  container=phaino-26122-1 job=ci-structured-merge-diff-benchmark
time="2020-02-14T20:55:55Z" level=info msg="Running command [\"go\" \"test\" \"-v\" \"-run='^$'\" \"-bench=.\" \"./...\"]..."

time="2020-02-14T20:57:41Z" level=fatal msg="Failed to write to log file \"$(ARTIFACTS)/benchmark-log.txt\"." error="open $(ARTIFACTS)/benchmark-log.txt: no such file or directory"
```

cjwagner commented on config/jobs/kubernetes-sigs/structured-merge-diff/structured-merge-diff-presubmits.yaml:
```diff
@@ -36,3 +36,21 @@ presubmits:
+    path_alias: sigs.k8s.io/structured-merge-diff
+    spec:
+      containers:
+      - image: gcr.io/k8s-prow/benchmarkjunit:latest
```
The image will be published to `gcr.io/k8s-testimages/` not `gcr.io/k8s-prow/`.

julianvmodesto replied:
Done: fixed the image name to the one pushed

cjwagner commented on config/jobs/kubernetes-sigs/structured-merge-diff/structured-merge-diff-presubmits.yaml:
```diff
@@ -36,3 +36,21 @@ presubmits:
+      containers:
+      - image: gcr.io/k8s-prow/benchmarkjunit:latest
+        command:
+        - /app/pkg/benchmarkjunit/app.binary
```
nit: You can just use `/benchmarkjunit` if you include something like this in the `prow_image` bazel target: https://github.com/kubernetes/test-infra/blob/fd45432efbe1eecb848d9a87127a61db4f2eeee3/prow/cmd/branchprotector/BUILD.bazel#L19-L20

julianvmodesto replied:
Done -- built the binary to `/benchmarkjunit` and set as the `command`

julianvmodesto commented:
Ahh makes sense. Added a postsubmit to publish the new image in a consistent manner, and updated (and tested) the new benchmark jobs accordingly.

julianvmodesto commented:
Fixed the cloudbuild and validated the push job with:

```
$ bazel run //images/builder -- --project=my-project --build-dir=. pkg/benchmarkjunit/
...
2020/02/14 17:48:44 Finished.
```

cjwagner commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -584,6 +584,32 @@ postsubmits:
+    branches:
+    - master
+    spec:
+      serviceAccountName: pusher
```
I think you need the `deployer` service account like the other image publishing jobs? The comment about using `pusher` instead is saying that we should switch to using `pusher`, but that involves transferring permissions that I don't think `pusher` has yet.

julianvmodesto replied:
Ah gotcha, fixed

cjwagner commented on pkg/benchmarkjunit/Dockerfile:
```diff
@@ -0,0 +1,32 @@
+
+FROM golang:1.13-alpine
+
+COPY --from=builder /app /benchmarkjunit
```
I'm not familiar with how Dockerfiles work with multiple `FROM`s, `COPY --from=`, or `FROM ... AS ...`, but this seems like overkill for wrapping a simple go binary in a container.
I think it would be simpler and clearer to build the go binary with a step in cloudbuild.yaml. Here is an example: https://github.com/kubernetes/test-infra/blob/7240a8ccbc6c8a56219b90e4292a9214797f1190/logexporter/cloudbuild.yaml#L2-L14

julianvmodesto replied:
Ah thanks for the example. Simplified!

julianvmodesto commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -584,6 +584,32 @@ postsubmits:
+      testgrid-dashboards: "sig-testing-images"
+      testgrid-tab-name: "benchmarkjunit"
+      testgrid-alert-email: kubernetes-sig-testing-alerts@googlegroups.com, k8s-infra-oncall@google.com
+      testgrid-num-failures-to-alert: '1'
```
Done, set to 3

julianvmodesto commented:
Addressed your comments!

And tested that both jobs work.

Cloud build job test:
```
$ bazel run //images/builder -- --project=my-project --build-dir=. pkg/benchmarkjunit/
...
2020/02/20 18:09:00 Running build jobs...
...
2020/02/20 18:10:14 Finished.
```

Benchmark job test:
```
$ bazel run //config:mkpj -- --job=ci-structured-merge-diff-benchmark > benchmark-ci-job.yaml
...

$ sed -i '/ARTIFACTS/d' benchmark-ci-job.yaml # phaino doesn't know about podutils

$ bazel run //prow/cmd/phaino -- $(pwd)/benchmark-ci-job.yaml
...
INFO[0001] Starting job...                               job=ci-structured-merge-diff-benchmark
INFO[0001] Waiting for job to finish...                  container=phaino-8470-1 job=ci-structured-merge-diff-benchmark
time="2020-02-21T17:58:00Z" level=info msg="Running command [\"go\" \"test\" \"-v\" \"-run='^$'\" \"-bench=.\" \"./...\"]..."
INFO[0098] PASS                                          duration=1m38.864825229s job=ci-structured-merge-diff-benchmark
INFO[0098] SUCCESS
```
